components/ui/data-table.tsx

```diff
@@ -0,0 +1,108 @@
+'use client'
+
+import {
+  ColumnDef,
+  flexRender,
+  Table as TanstackTable,
+} from '@tanstack/react-table'
+
+import {
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+  TableRow,
+} from '@/components/ui/table'
+
+import { ChevronLeft, ChevronRight, Loading } from '@/components/site/icons'
+
+import { Button } from './button'
+
+
+interface DataTableProps<TData, TValue> {
+  columns: ColumnDef<TData, TValue>[],
+  table: TanstackTable<TData>,
+  hasNextPage: boolean | undefined,
+  isFetching: boolean
+}
+
+export function DataTable<TData, TValue>({
+  columns,
+  table,
+  hasNextPage,
+  isFetching
+}: DataTableProps<TData, TValue>) {
+
+  return (
+    <>
+      <div className='flex items-center justify-end space-x-2 py-4'>
+        <Button
+          variant='outline'
+          size='sm'
+          onClick={() => table.previousPage()}
+          disabled={!table.getCanPreviousPage()}
+        >
+          <ChevronLeft className='h-4 w-5' />
+        </Button>
+        <div className='flex items-center gap-2'>
+          {table.getState().pagination.pageIndex}
+        </div>
+        <Button
+          variant='outline'
+          size='sm'
+          onClick={() => table.nextPage()}
+          disabled={!table.getCanNextPage() && (!hasNextPage || isFetching)}
+        >
+          { !table.getCanNextPage() && isFetching
+            ? <Loading className='h-5 w-5'/>
+            : <ChevronRight className='h-4 w-5' /> }
+        </Button>
+      </div>
+      <div className='rounded-md border'>
+        <Table>
+          <TableHeader>
+            {table.getHeaderGroups().map((headerGroup) => (
+              <TableRow key={headerGroup.id}>
+                {headerGroup.headers.map((header) => {
+                  return (
+                    <TableHead key={header.id}>
+                      {header.isPlaceholder
+                        ? null
+                        : flexRender(
+                          header.column.columnDef.header,
+                          header.getContext()
+                        )}
+                    </TableHead>
+                  )
+                })}
+              </TableRow>
+            ))}
+          </TableHeader>
+          <TableBody>
+            {table.getRowModel().rows?.length ? (
+              table.getRowModel().rows.map((row) => (
+                <TableRow
+                  key={row.id}
+                  data-state={row.getIsSelected() && 'selected'}
+                >
+                  {row.getVisibleCells().map((cell) => (
+                    <TableCell key={cell.id}>
+                      {flexRender(cell.column.columnDef.cell, cell.getContext())}
+                    </TableCell>
+                  ))}
+                </TableRow>
+              ))
+            ) : (
+              <TableRow>
+                <TableCell colSpan={columns.length} className='h-24 text-center'>
```

Comment:
Would one somewhat be able to retrieve the `columns.lenght` from inside of the `table` object?
'Cause if it is possible, we could remove the `columns` argument to be passed to this component

Reply:
As we saw earlier on that video, one can supply a `meta` argument on the `useReactTable` hook to introduce external context that are carried by the `table` else where it goes. 
It can be a solution for the `hasNextPage` and `isFetching`, for example